compiler/x/codegen/OMRMemoryReference.cpp

```diff
@@ -524,7 +524,7 @@ OMR::X86::MemoryReference::populateMemoryReference(
       evalSubTree = false;
 
    if (evalSubTree &&
-       subTree->getReferenceCount() > 1 || subTree->getRegister() != NULL || (self()->inUpcastingMode() && !subTree->cannotOverflow()))
+       (subTree->getReferenceCount() > 1 || subTree->getRegister() != NULL || (self()->inUpcastingMode() && !subTree->cannotOverflow())))
```

Comment:
hmm I'm not sure if this is right - the intent might have been to say if the refcount is greater than 1 and we want to eval OR if we have a register OR other condition since you won't eval in the other two cases... @0dvictor thoughts?

Reply:
I think the logic is correct, though it might not be quite obvious. It lists three conditions:
1. The reference count is greater than 1;
2. The tree has been evaluated;
3. CG is in upcasting mode (in the processing of evaluating `i2l`, `s2l`, `s2i`), AND the tree may overflow.

Obviously, we should evaluate the tree under case 1; otherwise we may incorrectly evaluate multiple times for the subsequent occurrences.
For case 2, since the tree has been evaluated, "re-evaluate" won't generate any instructions but simply returns its result register. The use of the register is same as if we freshly evaluate the tree.
For case 3, we need insert the proper sign-extension instruction so that we have to evaluate the tree regardless of its reference count.
So the condition looks correct to me.

Reply:
The issue I have is that evalSubTree is only true for some specific loadaddr so I'm trying to figure out why that is and why it should feed through as it is doing. I'm not saying the analysis above is wrong, but I'm not sure if there is an unintended change here (see 521 to 524 above).

Reply:
I see your point, and I agree three may be an unintended change. The original code is equivalent to
```
if (   (evalSubTree && 
        subTree->getReferenceCount() > 1)       || subTree->getRegister() != NULL || (self()->inUpcastingMode() && !subTree->cannotOverflow()))
```
`&&` has higher precedence than `||` so that `evalSubTree && subTree->getReferenceCount() > 1` is evaluated first.
Maybe we should move `subTree->getReferenceCount() > 1` into the earlier condition?

Reply:
I have fixed this one as I agree the earlier association of && makes more sense, especially given the archeology @andrewcraik went through to track back how this condition has evolved over the years (though we didn't go *all* the way back :) ). I also added a comment to capture that we're preserving the original code's behaviour and overriding the indentation that suggested a different intent.